src/Application/Services/UserService.cs

```diff
@@ -63,111 +63,150 @@ public async Task AddUserRole(User user, Role role, CancellationToken cancellati
         await _dbContext.SaveChangesAsync(cancellationToken);
     }
 
-    public int CreateUser(User user)
+    public User CreateUser(User user)
     {
         return CreateUser(user, CancellationToken.None).Result;
     }
 
-    public async Task<int> CreateUser(User newUser, CancellationToken cancellationToken)
+    public async Task<User> CreateUser(User newUser, CancellationToken cancellationToken)
     {
-        var currentUser = await _dbContext.Users
-            .TagWithContext("GetUserByEmail")
-            .Where(u => u.Email == newUser.Email)
-            .FirstOrDefaultAsync(cancellationToken);
+        try
+        {
+            if (string.IsNullOrWhiteSpace(newUser.Email))
+                throw new ArgumentException("Email is required", nameof(newUser));
+
+            var existingUser = await GetUserByEmailAsync(newUser.Email, cancellationToken);
+            if (existingUser != null)
+            {
+                _logger.LogInformation("User with email {Email} already exists, returning existing user", newUser.Email);
+                return existingUser;
+            }
 
-        if (currentUser != null)
+            _logger.LogInformation("Creating new user with email {Email}", newUser.Email);
+        
+            await AssignUserRolesAsync(newUser, cancellationToken);
+            await HandleStaffMemberCreationAsync(newUser, cancellationToken);
+            await ProcessUnclaimedAchievementsAsync(newUser, cancellationToken);
+
+            _dbContext.Users.Add(newUser);
+            await _dbContext.SaveChangesAsync(cancellationToken);
+
+            _cacheService.Remove(CacheTags.NewlyUserCreated);
+        
+            _logger.LogInformation("Successfully created user with ID {UserId}", newUser.Id);
+            return newUser;
+        }
+        catch (Exception ex)
         {
-            _logger.LogWarning("User with {email} already exists", newUser.Email);
-            return currentUser.Id;
+            _logger.LogError(ex, "Error creating user with email {Email}", newUser.Email);
+            throw;
         }
+    }
 
-        var userRole = await _dbContext.Roles
-            .TagWithContext("GetRoleNamedUser")
-            .FirstOrDefaultAsync(r => r.Name == "User");
-
-        newUser.Roles.Add(new UserRole { Role = userRole });
+    private async Task AssignUserRolesAsync(User newUser, CancellationToken cancellationToken)
+    {
+        var requiredRoles = new List<string> { "User" };
+        
+        if (IsStaffEmail(newUser.Email!))
+        {
+            requiredRoles.Add("Staff");
+        }
 
-        var userEmail = new MailAddress(newUser.Email);
+        var roles = await _dbContext.Roles
+            .TagWithContext("GetRequiredRoles")
+            .Where(r => requiredRoles.Contains(r.Name))
+            .ToListAsync(cancellationToken);
 
-        if (userEmail.Host == _staffSMTPDomain)
+        if (roles.Count != requiredRoles.Count)
```

Comment:
I would recommend something like this, to verify we have required roles rather than fail if the count is different.

```suggestion
        if (!requiredRoles.All(x => roles.Any(r => x.Id == r.Id)))
```

Reply:
Done - implement a check for all the values themselves.